File: src/main/java/com/github/t1/yaml/dump/Presenter.java
package com.github.t1.yaml.dump;

import com.github.t1.yaml.model.Alias;
import com.github.t1.yaml.model.Comment;
import com.github.t1.yaml.model.Directive;
import com.github.t1.yaml.model.Document;
import com.github.t1.yaml.model.Mapping;
import com.github.t1.yaml.model.Mapping.Entry;
import com.github.t1.yaml.model.Node;
import com.github.t1.yaml.model.Scalar;
import com.github.t1.yaml.model.Scalar.Line;
import com.github.t1.yaml.model.Sequence;
import com.github.t1.yaml.model.Sequence.Item;
import com.github.t1.yaml.model.Stream;
import com.github.t1.yaml.model.Visitor;
import lombok.RequiredArgsConstructor;

import static com.github.t1.yaml.tools.Tools.spaces;

@RequiredArgsConstructor public class Presenter {
    private static final char NL = '\n';

    private final StringBuilder out = new StringBuilder();
    private Visitor visitor = new Visitor() {
        boolean skipNextIndent;
        boolean skipIndent;
        int indent = 0;

        private String indent() {
            if (skipIndent)
                return "";
            if (skipNextIndent) {
                skipNextIndent = false;
                return "";
            }
            return spaces(indent * 2);
        }

        @Override public void visit(Directive directive) { out.append("%").append(directive.name()).append(" ").append(directive.parameters()).append(NL); }

        @Override public void leaveDirectives(Document document) { out.append("---").append(NL); }

        @Override public void visitPrefixComment(Comment prefixComment) { append(prefixComment).append(NL); }

        @Override public void leaveBody(Node node) { nl(); }

        @Override public void enterDocumentEnd() { out.append("..."); }

        @Override public void visitSuffixCommend(Comment suffixComment) { append(suffixComment); }

        @Override public void leaveDocumentEnd() { nl(); }


        @Override public void visit(Alias alias) { out.append(alias); }


        @Override public void enterSequenceItem(Sequence sequence, Item item) {
            switch (sequence.style()) {
                case FLOW:
                    skipIndent = true;
                    out.append((item == sequence.firstItem()) ? "[" : ", ");
                    break;
                case BLOCK:
                    out.append(indent());
                    skipNextIndent = !item.nl();
                    indent++;
                    out.append("-").append(item.nl() ? "\n" : " ");
            }
        }

        @Override public void leaveSequenceItem(Sequence sequence, Item item) {
            switch (sequence.style()) {
                case FLOW:
                    if (item == sequence.lastItem())
                        out.append("]");
                    skipIndent = false;
                    break;
                case BLOCK:
                    if (item != sequence.lastItem())
                        nl();
                    indent--;
            }
        }


        @Override public void visit(Scalar scalar) {
            if (scalar.tag() != null)
                out.append(scalar.tag()).append(' ');
            out.append(scalar.style().quote());
        }

        @Override public void visit(Line line) {
            out.append(indent()).append(spaces(line.indent()));
            out.append(line.text());
            if (line.comment() != null)
                append(line.comment());
        }

        @Override public void leaveScalarLine(Scalar node, Line line) {
            if (line != node.lastLine())
                nl();
        }

        @Override public void leave(Scalar scalar) { out.append(scalar.style().quote()); }


        @Override public void enterMappingKey(Entry entry, Node key) { out.append(entry.hasMarkedKey() ? "? " : ""); }

        @Override public void enterMappingValue(Entry entry, Node key) {
            indent++;
            skipNextIndent = !entry.hasNlAfterKey();
            out.append(":").append(entry.hasNlAfterKey() ? NL : ' ');
        }

        @Override public void leaveMappingValue(Entry entry, Node key) { indent--; }

        @Override public void leaveMappingEntry(Mapping node, Entry entry) {
            if (entry != node.lastEntry())
                nl();
        }

        private StringBuilder append(Comment comment) {
            return out.append(spaces(comment.indent())).append("# ").append(comment.text());
        }

        private void nl() { out.append(NL); }
    };

    public String present(Stream stream) {
        stream.guide(visitor);
        return out.toString();
    }
}

File: src/main/java/com/github/t1/yaml/dump/Presenter.kt
package com.github.t1.yaml.dump

import com.github.t1.yaml.model.Alias
import com.github.t1.yaml.model.Collection.Style.BLOCK
import com.github.t1.yaml.model.Collection.Style.FLOW
import com.github.t1.yaml.model.Comment
import com.github.t1.yaml.model.Directive
import com.github.t1.yaml.model.Document
import com.github.t1.yaml.model.Mapping
import com.github.t1.yaml.model.Mapping.Entry
import com.github.t1.yaml.model.Node
import com.github.t1.yaml.model.Scalar
import com.github.t1.yaml.model.Scalar.Line
import com.github.t1.yaml.model.Scalar.Style.DOUBLE_QUOTED
import com.github.t1.yaml.model.Scalar.Style.PLAIN
import com.github.t1.yaml.model.Scalar.Style.SINGLE_QUOTED
import com.github.t1.yaml.model.Sequence
import com.github.t1.yaml.model.Sequence.Item
import com.github.t1.yaml.model.Stream
import com.github.t1.yaml.model.Visitor
import com.github.t1.yaml.tools.spaces

class Presenter {
    private val out = StringBuilder()
    private val visitor = object : Visitor {
        var skipNextIndent: Boolean = false
        var skipIndent: Boolean = false
        var indent = 0

        private fun indent(): String {
            if (skipIndent)
                return ""
            if (skipNextIndent) {
                skipNextIndent = false
                return ""
            }
            return spaces(indent * 2)
        }

        override fun visit(directive: Directive) {
            out.append("%").append(directive.name).append(" ").append(directive.parameters).append(NL)
        }

        override fun leaveDirectives(document: Document) {
            out.append("---").append(NL)
        }

        override fun visitPrefixComment(prefixComment: Comment) {
            append(prefixComment).append(NL)
        }

        override fun leaveBody(node: Node) {
            nl()
        }

        override fun enterDocumentEnd() {
            out.append("...")
        }

        override fun visitSuffixCommend(suffixComment: Comment) {
            append(suffixComment)
        }

        override fun leaveDocumentEnd() {
            nl()
        }


        override fun visit(alias: Alias) {
            out.append(alias)
        }


        override fun visit(scalar: Scalar) {
            if (scalar.tag != null)
                out.append(scalar.tag).append(' ')
            out.append(scalar.style.quote)
        }

        override fun enterScalarLine(node: Scalar, line: Line) {
            out.append(indent()).append(spaces(line.indent))
            out.append(when (node.style) {
                PLAIN -> line.text
                SINGLE_QUOTED -> line.text.replace("'", "''")
                DOUBLE_QUOTED -> line.text
            })
        }

        override fun leaveScalarLine(scalar: Scalar, line: Line) {
            out.append(scalar.style.quote)
            if (line.comment != null)
                append(line.comment!!)
            if (line != scalar.lastLine)
                nl()
        }


        override fun enterSequenceItem(sequence: Sequence, item: Item) {
            when (sequence.style) {
                FLOW -> {
                    skipIndent = true
                    out.append(if (item === sequence.firstItem) "[" else ", ")
                }
                BLOCK -> {
                    out.append(indent())
                    skipNextIndent = !item.nl
                    indent++
                    out.append("-").append(if (item.nl) "\n" else " ")
                }
            }
        }

        override fun leaveSequenceItem(sequence: Sequence, item: Item) {
            when (sequence.style) {
                FLOW -> {
                    if (item === sequence.lastItem)
                        out.append("]")
                    skipIndent = false
                }
                BLOCK -> {
                    if (item !== sequence.lastItem)
                        nl()
                    indent--
                }
            }
        }


        override fun enterMappingEntry(mapping: Mapping, entry: Entry) {
            when (mapping.style) {
                FLOW -> out.append(if (entry === mapping.firstEntry) "{" else ",")
                BLOCK -> {
                }
            }
        }

        override fun enterMappingKey(entry: Entry, key: Node) {
            out.append(if (entry.hasMarkedKey) "? " else "")
        }

        override fun enterMappingValue(entry: Entry, key: Node) {
            indent++
            skipNextIndent = !entry.hasNlAfterKey
            out.append(":").append(if (entry.hasNlAfterKey) NL else ' ')
        }

        override fun leaveMappingValue(entry: Entry, key: Node) {
            indent--
        }

        override fun leaveMappingEntry(mapping: Mapping, entry: Entry) {
            when (mapping.style) {
                FLOW -> if (entry === mapping.lastEntry) out.append("}")
                BLOCK -> if (entry !== mapping.lastEntry) nl()
            }
        }

        private fun append(comment: Comment): StringBuilder {
            return out.append(spaces(comment.indent)).append("# ").append(comment.text)
        }

        private fun nl() {
            out.append(NL)
        }
    }

    fun present(stream: Stream): String {
        stream.guide(visitor)
        return out.toString()
    }

    companion object {
        private const val NL = '\n'
    }
}
